activerecord/lib/active_record/relation/predicate_builder.rb

```diff
@@ -145,5 +150,9 @@ def can_be_bound?(column_name, value)
         handler_for(value).is_a?(BasicObjectHandler) &&
         !table.associated_with?(column_name)
     end
+
+    def polymorphic_array?(value)
+      value.is_a?(Array) && value.map(&:class).map { |e| e.try(:base_class) }.uniq.size > 1
```

Comment:
This line definitely can be more efficient. This also needs to return immediately if the first element of the array isn't an Active Record object, as this has introduced a > linear time performance regression whenever an array is passed to `where`